File: Find_Eventual_Safe_States.cpp
class Solution {
public:
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        int n = graph.size();
        std::vector<int> visited(n, 0); // 0: unvisited, 1: visiting, 2: terminal
        std::vector<int> result;

        for (int i = 0; i < n; i++) {
            if (isSafe(graph, visited, i)) {
                result.push_back(i);
            }
        }

        return result;        
    }

private:
    bool isSafe(std::vector<std::vector<int>>& graph, std::vector<int>& visited, int node) {
        if (visited[node] > 0) {
            return visited[node] == 2;
        }

        visited[node] = 1;
        for (int neighbor : graph[node]) {
            if (!isSafe(graph, visited, neighbor)) {
                return false;
            }
        }

        visited[node] = 2;
        return true;
    }
};

File: Find_Eventual_Safe_States.py
class Solution:
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        n = len(graph)
        visited = [0] * n # 0: unvisited, 1: visiting, 2: terminal
        result = []

        def isSafe(node):
            if visited[node] > 0:
                return visited[node] == 2
            
            visited[node] = 1
            for neighbor in graph[node]:
                if not isSafe(neighbor):
                    return False
            
            visited[node] = 2
            return True

        for i in range(n):
            if isSafe(i):
                result.append(i)
        
        return result
